src/task_manager.cc

```diff
@@ -47,12 +47,27 @@ namespace detail {
 		// This method is called from the executor thread, but does not lock task_mutex to avoid lock-step execution with the main thread.
 		// latest_horizon_reached does not need synchronization (see definition), all other accesses are implicitly synchronized.
 
-		assert(get_task(epoch_tid)->get_type() == task_type::epoch);
+		const auto tsk = get_task(epoch_tid);
+
+		assert(tsk->get_type() == task_type::epoch);
 		assert(!m_latest_horizon_reached || *m_latest_horizon_reached < epoch_tid);
 		assert(m_latest_epoch_reached.get() < epoch_tid);
 
 		m_latest_epoch_reached.set(epoch_tid);
 		m_latest_horizon_reached = std::nullopt; // Any non-applied horizon is now behind the epoch and will therefore never become an epoch itself
+
+		if(tsk->get_epoch_action() == epoch_action::barrier) {
+			// Wait for the main thread to call resume_after_barrier. This ensures that no commands can be executed before all captures have been exfiltrated.
+			// It is not sufficient for the main thread to stop submitting work, since e.g. AWAIT PUSH commands do not depend on a task definition.
```

Comment:
Ahh okay I think I get it now. I think I was confused because I thought there should be an anti-dependency from an await push onto the exfiltrating epoch that prevents this race... And there is, but that's not enough because the exfiltration happens in the main thread, not the epoch worker job. So we need a way for that job to wait until the main thread is done exfiltrating. Correct?